### PR TITLE
Flush / Stats / Key Listings: backfilled missing methods

### PR DESCRIPTION
**Pull request by bryanlatten**
Added `getAllKeys` and `getStats` and `flush`

### CODE REVIEW

File: src/Adapters/MemcacheAdapter.php [hunk 1]
```diff
@@ -205,6 +209,76 @@ protected function _deleteMulti( array $keys ) {
...
+
+      $slab_ids = array_keys( $slabs );
+
+      foreach ( $slab_ids as $slab_id ) {
+
+        if ( !is_integer( $slab_id ) ) {
+          continue;
+        }
+
+        $cache_dump = $connection->getExtendedStats( self::STAT_KEY_DUMP, (int)$slab_id );
```
Comment (ingluisjimenez) [inline src/Adapters/MemcacheAdapter.php]: probably don't need to cast to int, as there is_integer() above will enforce an integer to get to this point

File: src/Adapters/MemcacheAdapter.php [hunk 2]
```diff
@@ -205,6 +209,76 @@ protected function _deleteMulti( array $keys ) {
...
+
+        foreach ( array_values( $cache_dump ) as $entries ) {
+
+          if ( is_array( $entries ) ) {
```
Comment (ingluisjimenez) [inline src/Adapters/MemcacheAdapter.php]: invert condition and `continue;`

File: src/Factory.php [hunk 3]
```diff
@@ -34,4 +36,32 @@ public static function create( ConfigService $config, $type, EventDispatcherInte
 
   } // create
 
+
+  /**
+   * @throws Behance\NBD\Cache\Exceptions\SystemRequirementException
```
Comment (ingluisjimenez) [inline src/Factory.php]: missing reason: `If ....`

**Comment by ingluisjimenez** [comment]
:+1:

File: README.md [hunk 4]
```diff
@@ -40,6 +40,9 @@ Methods
 <tr><td>decrement( $key, $value = 1 )</td><td>Decrements $key by $value</td></tr>
 <tr><td>delete( $key )</td><td>Removes a single key from server</td></tr>
 <tr><td>deleteMulti( array $keys )</td><td>Removes group of keys from server(s)</td></tr>
+<tr><td>getAllKeys()</td><td>Retrieves the full keylist from server(s)</td></tr>
+<tr><td>getStats()</td><td>Retrieves usage stats from server(s)</td></tr>
+<tr><td>bind( $event_name, callable $handler )</td><td>Provide handlers for specific cache-specific events</td></tr>
```
Comment (mlebrun) [inline README.md]: `specific cache-specific`?

File: src/Adapters/MemcacheAdapter.php [hunk 5]
```diff
@@ -205,6 +209,76 @@ protected function _deleteMulti( array $keys ) {
...
+
+    $results      = [];
+    $connection   = $this->_connection;
+    $server_slabs = $connection->getExtendedStats( self::STAT_KEY_SLABS );
+
+    if ( empty( $server_slabs ) ) {
+      return $results;
+    }
+
+    foreach ( array_values( $server_slabs ) as $slabs ) {
```
Comment (mlebrun) [inline src/Adapters/MemcacheAdapter.php]: do you need to call values on it?

File: src/Adapters/MemcacheAdapter.php [hunk 6]
```diff
@@ -205,6 +209,76 @@ protected function _deleteMulti( array $keys ) {
...
+
+      foreach ( $slab_ids as $slab_id ) {
+
+        if ( !is_integer( $slab_id ) ) {
+          continue;
+        }
+
+        $cache_dump = $connection->getExtendedStats( self::STAT_KEY_DUMP, $slab_id );
+
+        foreach ( array_values( $cache_dump ) as $entries ) {
```
Comment (mlebrun) [inline src/Adapters/MemcacheAdapter.php]: and here

File: src/Factory.php [hunk 7]
```diff
@@ -13,19 +14,20 @@ class Factory {
...
   const TYPE_MEMCACHED     = 'Memcached';
 
+
   /**
    * TODO: auto-choose type based on available extensions
    *
    * @param Behance\NBD\Cache\Services\ConfigService $config
-   * @param string $type
    * @param Symfony\Component\EventDispatcher\EventDispatcherInterface $event_dispatcher
+   * @param string $type  which kind of adapter to build, if omitted will choose automatically
```
Comment (mlebrun) [inline src/Factory.php]: you moved this after the dispatcher, but it doesn't look like that was reflected in the constructor

**Comment by mlebrun** [comment]
:+1:

**Pull request merged**